### PR TITLE
[diem framework] delete tiered minting logic

### PR DESCRIPTION
**Pull request by sblackshear**
Tiered minting is no longer a requirement, so this PR eliminates all logic related to tier restrictions.

The `tier_index` parameter to the `tiered_mint` public function + transaction script remain for backward compatibility, but will now be ignored.
In addition, `tiered_mint<T>(dd_address, ...)` will now delete the `TierInfo<T>` resource under `dd_address`. We can delete the `TierInfo` resource altogether once there are no longer any left in the system.

### CODE REVIEW

File: language/tools/vm-genesis/src/lib.rs
Comment (sblackshear) [inline language/tools/vm-genesis/src/lib.rs:314]: @runtian-zhou: Is deleting this ok?

File: language/diem-framework/modules/DesignatedDealer.move
Comment (moeziniaf) [inline language/diem-framework/modules/DesignatedDealer.move:59]: Are we OK to reuse move error codes going forward? Or are we treating them as part of our compatibility set?

File: language/diem-framework/modules/DesignatedDealer.move
Comment (moeziniaf) [inline language/diem-framework/modules/DesignatedDealer.move:135]: Just checking we do indeed want to leave `TierInfo` here even after the first call that resource won't exist (but we need it for the first call)?

File: language/diem-framework/modules/DesignatedDealer.move
Comment (tzakian) [reply language/diem-framework/modules/DesignatedDealer.move:59]: Good question...I am of the opinion that we can reuse these as the Move explain tool will pickup these changes. Additionally, only the error categories are guaranteed to be stable, whereas the reasons (these numbers) don't have the same stability guarantees. Interested in @sblackshear's  thoughts on this as this is the first real case where we've deleted a lot of error reasons.

**Review by tzakian — COMMENTED** [review]
This looks good to me (and love the amount of code deleted!). But I leave final approval up to @DavidLDill and @emmazzz as I know they've both been looking at some verification issues relating to tiered mint and want to make sure they get a chance to look at this before it lands.

File: language/diem-framework/modules/DesignatedDealer.move
Comment (sblackshear) [reply language/diem-framework/modules/DesignatedDealer.move:135]: Yes, the bytecode verifier forces us to have `acquires TierInfo` anywhere that there's a `move_from<TierInfo>`.

File: language/diem-framework/modules/DesignatedDealer.move
Comment (sblackshear) [reply language/diem-framework/modules/DesignatedDealer.move:59]: Yes, very good question... I think reusing seems harmful and we should avoid it. In particular, I'm thinking of a case where a client has an SDK with the old error codes, but we have changed the meaning of (e.g.) error code 1 in the current version of the framework. For example, if 1 used to mean "insufficient funds" and it now means "recipient does not exist", the client with the old SDK may react to the error code in the wrong way.

In terms of what we should do in this PR, I guess what needs to be decided is a convention for "reserving" old error codes that were previously used and should not be reused in the future:
- Comment them out + note that they are deprecated. By convention, new error codes will always be added to the "end" (I think this was already the case)
- Leave them in, but change the names so they will no longer show up as error codes for `move explain` (e.g., `DEPRECATED_EINVALID_TIER_ADDITION`

I have a slight bias for the first one, particularly if we are going to expose constants in the transaction builders (this has been recently requested for `Roles`). We could have the transaction builder ignore constants named `DEPRECATED_*`, but that doesn't seem nice. Thoughts? Alternatives?

File: language/tools/vm-genesis/src/lib.rs
Comment (runtian-zhou) [reply language/tools/vm-genesis/src/lib.rs:314]: Deleting this means that we won't test this logic in language_e2e_tests, which I think is OK. Because it would be deprecated code.

**Review by DavidLDill — APPROVED** [review]
Specs look ok to me.  If the Prover is happy with it, I'm happy.

File: language/diem-framework/modules/DesignatedDealer.move
Comment (moeziniaf) [reply language/diem-framework/modules/DesignatedDealer.move:59]: I have the same opinion - also, these error codes will be dead code so they won't be used by clients in any case? commenting them is better imo as I would class changing codes as a breaking change

**Pull request closed without merging**
